runtime/signals_nat.c

```diff
@@ -85,3 +85,37 @@ void caml_garbage_collection(void)
                               nallocs, alloc_len);
   }
 }
+
+/* Trap handling for the POWER architecture.  Convert the trap into
+   an out-of-bounds exception. */
+
+#if defined(TARGET_power)
+
+extern void caml_array_bound_error_asm(void);
+
+void caml_sigtrap_handler(int signo, siginfo_t * info, void * context)
+{
+  /* The trap occurs in ocamlopt-generated code. */
+  /* The purpose of this function is to simulate a [caml_c_call]
+     to [caml_array_bound_error_asm]. */
+
+  caml_domain_state * dom_st = Caml_state;
```

Comment:
Nice, but:
- is it known that the signal handler will only run for OCaml code? I wonder if it is necessary to check that the pc points to OCaml code prior to accessing `Caml_state`.
- For C11 conformance, one can access TLS in a signal handler, but the thread-local variable has to be `_Atomic`, which it is not, currently. One also has to consider the fact that glibc does not conform to C11 and can allocate in corner cases (that are probably irrelevant here); as for MacOS one might want to reimplement the more efficient implementation of `Caml_state` that uses explicit TLS one day. Could you side-step these worries by retrieving the domain state pointer from `ctx` instead?

Reply:
Where can I find the code in ocamlopt that is setting up the SIGTRAP machinery? 
Is `runtime/signals.c` the right place?

Does the Reraise sigsegv #10409 need multicore updating? @gadmm

Reply:
@tmcgilchrist It makes sense to adapt #10409 to SIGTRAP on Power indeed (as part of a new PR and if someone volunteers).

Reply:
> Could you side-step these worries by retrieving the domain state pointer from ctx instead?

Yes, sure,  see 0b12d861c075891eaccac993f500b2b990060c73 .

Reply:
> is it known that the signal handler will only run for OCaml code? I wonder if it is necessary to check that the pc points to OCaml code prior to accessing Caml_state.

And what shall we do if it doesn't?  Abort the program?

Reply:
For SEGV you used to restore the signal handler to its default disposition and restart at the faulting instruction (and thus abort the program in a debuggable way). I imagine that you did not do the same here for some reason. I imagine that traps increment the program counter like on x86 but decrementing the pc does not sound like an obstacle. Am I missing anything?

I also noticed that `caml_find_code_fragment_by_pc` is essentially signal-safe because the call to `caml_code_fragment_cleanup` in `cycle_all_domains_callback` can be omitted in native, as no code fragment gets removed.

I agree with your point that none of these issues are blocking, since this is old code being ported to multicore.